src/java.base/share/native/libjli/java.c

```diff
@@ -568,40 +471,26 @@ JavaMain(void* _args)
 
     ret = 1;
 
-    /*
-     * See bugid 5030265.  The Main-Class name has already been parsed
-     * from the manifest, but not parsed properly for UTF-8 support.
-     * Hence the code here ignores the value previously extracted and
-     * uses the pre-existing code to reextract the value.  This is
-     * possibly an end of release cycle expedient.  However, it has
-     * also been discovered that passing some character sets through
-     * the environment has "strange" behavior on some variants of
-     * Windows.  Hence, maybe the manifest parsing code local to the
-     * launcher should never be enhanced.
-     *
-     * Hence, future work should either:
-     *     1)   Correct the local parsing code and verify that the
-     *          Main-Class attribute gets properly passed through
-     *          all environments,
-     *     2)   Remove the vestages of maintaining main_class through
-     *          the environment (and remove these comments).
-     *
-     * This method also correctly handles launching existing JavaFX
-     * applications that may or may not have a Main-Class manifest entry.
-     */
-    mainClass = LoadMainClass(env, mode, what);
-    CHECK_EXCEPTION_NULL_LEAVE(mainClass);
+    const jobject mainEntry = LocateMainEntry(env, mode, what);
+    CHECK_EXCEPTION_NULL_LEAVE(mainEntry);
+    const jclass mainEntryType = (*env)->GetObjectClass(env, mainEntry); // sun/launcher/LauncherHelper$MainEntry
+    const jfieldID mainClassFid = (*env)->GetFieldID(env, mainEntryType, "mainClass", "Ljava/lang/Class;");
```

Comment:
Using record fields directly in JNI feels risky. Can we switch to use the getters instead, even though access control is absent in JNI?